src/mkdocs_autorefs/references.py

```diff
@@ -197,13 +202,78 @@ def fix_refs(html: str, url_mapper: Callable[[str], str]) -> tuple[str, list[str
     return html, unmapped
 
 
+class AnchorScannerTreeProcessor(Treeprocessor):
+    """Tree processor to scan and register HTML anchors."""
+
+    _htags: ClassVar[set[str]] = {"h1", "h2", "h3", "h4", "h5", "h6"}
+
+    def __init__(self, plugin: AutorefsPlugin, md: Markdown | None = None) -> None:
+        """Initialize the tree processor.
+
+        Parameters:
+            plugin: A reference to the autorefs plugin, to use its `register_anchor` method.
+        """
+        super().__init__(md)
+        self.plugin = plugin
+        self._slug = md.treeprocessors["toc"].slugify
+
+    def run(self, root: Element) -> None:  # noqa: D102
+        if self.plugin.current_page is not None:
+            self._scan_anchors(root)
+
+    def _scan_anchors(self, parent: Element) -> list[str]:
+        ids = []
+        # We iterate on pairs of elements, to check if the next element is a heading (alias feature).
+        for el, next_el in zip_longest(parent, parent[1:], fillvalue=Element("/")):
+            if el.tag == "a":
+                # We found an anchor. Record its id if it has one.
+                if anchor_id := el.get("id"):
+                    if el.tail and el.tail.strip():
+                        # If the anchor has a non-whitespace-only tail, it's not an alias:
+                        # register it immediately.
+                        self.plugin.register_anchor(self.plugin.current_page, anchor_id)  # type: ignore[arg-type]
+                    else:
+                        # Else record its id and continue.
+                        ids.append(anchor_id)
+            elif el.tag == "p":
+                if ids := self._scan_anchors(el):
+                    # Markdown anchors are always rendered as `a` tags within a `p` tag.
+                    # Headings therefore appear after the `p` tag. Here the current element
+                    # is a `p` tag and it contains at least one anchor with an id.
+                    # We can check if the next element is a heading, and use its id as href.
+                    href = (next_el.get("id") or self._slug(next_el.text or "")) if next_el.tag in self._htags else ""
+                    for anchor_id in ids:
+                        self.plugin.register_anchor(self.plugin.current_page, anchor_id, href)  # type: ignore[arg-type]
+                    ids.clear()
+            else:
+                # Recurse into sub-elements.
+                ids = self._scan_anchors(el)
+        return ids
```

Comment:
Mm I think this implementation has some edge cases, I definitely found one like this

```markdown
            [](){#alias5}
            [](){#alias6}
            Decoy.
            ## Heading more
```
which produces
```python
'alias5': 'page#heading-more',
'alias6': 'page#alias6',
```

Probably it won't work to only have a lookahead of 1 tag ahead, I think you need indefinite lookahead. But there's a way to achieve that by instead thinking of a "buffer of pending anchors which may or may not become aliased to some heading", and flushing it whenever a non-anchor element appears. I will propose a modified implementation that does this.

Reply:
Awesome, thanks a lot! Cherry-picked and pushed :slightly_smiling_face: